server functions tested

limxuanyu127 opened the pull request:
- 'server2' is the new branch, 'server' branch is deprecated (leaving it there for a while just in case)
- Tests in BookingManagerTest should work
- Function params have changed, will update the design doc soon


Note on functions:
- Note the use of java.time library to manipulate time stuff
- queryAvailability returns an ArrayList of ArrayList<startTime, EndTime>
- facils have new functions (addBooking, removebooking etc.), that are called by BookingManager

Note on in-mem 'persistence':
- All data is stored in facilTable (hashtable), (key: facilName, value: facility object).
- Each facility object has a bookingsTable (hashtable), (key:date, value: ArrayList<booking>)

limxuanyu127 committed:
server functions tested, less observer/callback

limxuanyu127 committed:
server functions tested, less observer/callback

limxuanyu127 committed:
server functions tested, less observer/callback

limxuanyu127 committed:
Update server code to fit day String

limxuanyu127 committed:
Update server code to fit day String

limxuanyu127 committed:
Update server code to fit day String

sith0008 commented on src/main/java/server/entities/Facility.java:
```diff
@@ -19,45 +18,37 @@ public Facility(String name){
+        LocalTime endTime = booking.getEnd();
+        String day =booking.getDay();
+
         if((startTime.compareTo(BookingManager.getOpenTime()) <0) || endTime.compareTo(BookingManager.getCloseTime()) > 0){
             System.out.println("Booking invalid: start time is too early, or end time is too late");
             return;
         }
```
suggest to split into two clauses, so the client knows exactly what is the problem (too early or too late)

should we also return exception to BookingManager? jus like how we're doing it for `removeBooking`

limxuanyu127 replied:
good point, i will update this

limxuanyu127 replied:
edited such that Facility functions (addBooking and removeBooking) will not return exceptions. BookingManager will manage all exceptions created and returned. addBooking and removeBooking, will be put in a try catch block instead to catch unexpected errors, but this catch block should never be reached bcos BookingManager should have done the checks

sith0008 commented on src/main/java/server/managers/BookingManager.java:
```diff
@@ -19,40 +19,39 @@ public static LocalTime getCloseTime() {
             return new Pair<ArrayList, Exception>(results, e);
         }
 
         for (int j=0; j < dates.size();j++){
```
small tip: can do

```
for (LocalDate d : dates) {
    // use d, instead of dates.get()
}
```

sith0008 commented on src/main/java/server/managers/BookingManager.java:
```diff
@@ -61,67 +60,275 @@ public static LocalTime getCloseTime() {
+    public Pair<Booking, Exception> offsetBooking(String facilName, int bookingId, int offset, Hashtable facilTable){
+        //FIXME offset should be in minutes
+        Exception e;
 
-        facilName = facilName.toLowerCase();
+        Booking b = this.findFacilBookingById(facilName, bookingId, facilTable);
+        String day = b.getDay();
 
-        if (!facilTable.containsKey(facilName)){
-            Exception e = new NoSuchElementException("Facility does not exist"); //TODO review this exception
+        LocalTime newStart = b.getStart().plusMinutes(offset);
+        LocalTime newEnd = b.getEnd().plusMinutes(offset);
```
in the client side, i designed offset to be in slots instead of minutes. but either way works.

limxuanyu127 replied:
yeap i noticed, this is one of the things i will coordinate w xq when we integrating our parts

sith0008 commented on src/main/java/server/managers/BookingManager.java:
```diff
@@ -61,67 +60,275 @@ public static LocalTime getCloseTime() {
+        return new Pair<Booking, Exception>(b, null);
+
+    }
+    public Pair<Booking, Exception> extendBooking(String facilName, int bookingId, int offset, Hashtable facilTable){
+        //FIXME offset should be in minutes
+        Exception e;
+        Booking b = findFacilBookingById(facilName, bookingId, facilTable);
+        LocalTime bStart = b.getStart();
+        LocalTime bEnd = b.getEnd();
+        LocalTime newEnd = bEnd.plusMinutes(offset);
+        String day = b.getDay();
+
+        e = doBookingNotNullCheck(b);
+        if(e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+        e = doBookingCheck(facilName,bStart,newEnd,facilTable);
+        if (e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+        e = doAvailabilityCheckExceptCurrent(day, bookingId, facilName,bStart,newEnd,facilTable);
+        if (e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+
+        b.setEnd(newEnd);
+        return new Pair<Booking, Exception>(b, null);
+    }
+
+
```
not an issue but just curious, how come Facility does not provide support for extend and offset booking? i.e. for create, query and delete, we use Facility's methods, but not for extend and offset

limxuanyu127 replied:
wa good point, i think by my intended design (ie. Bookingmanager talks to Facility, Facility talks to Bookings/bookingslist) it should be what u mentioned.

sith0008 reviewed (commented):
realised it's a lot of work on the server side haha, sick stuff!!

one general comment:
- should we be returning exceptions? or returning objects and throwing exceptions? tho i dont really see a huge problem with the current impl, just wondering what's the best way to handle exceptions

xqlim1996 reviewed (approved):
awaiting testing for communicator integration

limxuanyu127 committed:
update

limxuanyu127 committed:
update

limxuanyu127 committed:
update

xqlim1996 committed:
New Server class

shifted while loop (for receiving) to Server class from ServerCommunicator class

xqlim1996 committed:
New Server class

shifted while loop (for receiving) to Server class from ServerCommunicator class

limxuanyu127 committed:
checkpoint

limxuanyu127 committed:
checkpoint

xqlim1996 committed:
Merge branch 'server2' of https://github.com/limxuanyu127/facility-booking into server2

xqlim1996 committed:
Merge branch 'server2' into ServerWrapper

xqlim1996 committed:
used xuanyu's impl

xqlim1996 committed:
used xuanyu's impl

xqlim1996 committed:
Setup AtMostOnce Experiment

xqlim1996 committed:
Setup AtMostOnce Experiment

xqlim1996 committed:
Merge branch 'main' into ServerWrapper

xqlim1996 committed:
Setup Experiments for Report

Lost Request, Lost Response (Idempotent), Lost Response (Non-Idempotent)

To run, run the server test then run the client test immediately

xqlim1996 committed:
Setup Experiments for Report

Lost Request, Lost Response (Idempotent), Lost Response (Non-Idempotent)

To run, run the server test then run the client test immediately

limxuanyu127 committed:
halfway through translator

xqlim1996 committed:
Merge branch 'server2' of https://github.com/limxuanyu127/facility-booking into server2

xqlim1996 committed:
Update translator.java

xqlim1996 committed:
small fix for translator

xqlim1996 committed:
Merge branch 'ServerWrapper' of https://github.com/limxuanyu127/facility-booking into ServerWrapper

xqlim1996 committed:
Create Translator.java~merged

xqlim1996 committed:
removed small translator

xqlim1996 committed:
Merge pull request #12 from limxuanyu127/ServerWrapper

Experiments for Report

limxuanyu127 committed:
finished translator, added communications to server which needs to be tested and cleaned up

limxuanyu127 committed:
clean up booking entity

limxuanyu127 committed:
edit

xqlim1996 committed:
Added Client IP and Port to AddObserver function

xqlim1996 commented on src/main/java/server/Server.java:
```diff
@@ -0,0 +1,157 @@
+            Exception e = null;
+
+            //TODO
+            // Run tests to see if type castings (eg. BookingFacilityResponse to Response) is valid
```
Don't think theres a need to cast since it extends Response

limxuanyu127 replied:
i see, yea my IDE keep showing me redlines so i thought i needed to cast, but yea i can take it out ltr

xqlim1996 commented on src/main/java/server/Server.java:
```diff
@@ -0,0 +1,157 @@
+            //TODO
+            // Run tests to see if type castings (eg. BookingFacilityResponse to Response) is valid
+            // For translator -> bookingManager + entities, there are some that still takes in client id, but those should be removed
+            // Update translator.notifyObservers() function to send notifications
```
The current implementation. should work but requires testing

xqlim1996 commented on src/main/java/server/translator/Translator.java:
```diff
@@ -0,0 +1,279 @@
+        }
+        else{
+            responseMessage = new ResponseMessage(200, "success") ;
+            ArrayList convertedList =localToDatetime(outputBooking);
```
Serializer has java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0 error for this list

sith0008 commented on src/main/java/server/entities/Booking.java:
```diff
@@ -9,65 +9,57 @@
+    public Booking(String day, int bookingId, String facilityName, LocalTime start, LocalTime end) {
+        this.day = day;
         this.bookingId = bookingId;
-        this.clientId = clientId;
```
not sure if it's important, but not storing clientID in booking means any client can edit/delete any booking by any client. I.e. do we want to validate client before allowing offset/extend/delete requests?

limxuanyu127 replied:
i see, i think can disucss this later. Bcos validating by IP and port will mean that a client has to use the same machine when he wants to access the services, so im not sure if we can use IP and port as client Id, or if not what better ways are there to validate.

sith0008 replied:
okay, we can discuss. i think validation might not be an important thing to do for the scope of this project bah haha. but i was thinking IP+port represents userID or like session token in an actual scenario

sith0008 commented on src/main/java/server/entities/Facility.java:
```diff
@@ -19,78 +18,114 @@ public Facility(String name){
-        return name;
+    private Boolean isValidObserver(FacilityObserver o){
+
+        if (o.getEndDate().compareTo(LocalDateTime.now()) <0 ){
+            return false;
+        }
+        else{
+            return true;
+        }
+
     }
```
not an issue, just a tip for code cleanliness: 

`return (o.getEndDate().compareTo(LocalDateTime.now()) >= 0 )`

sith0008 commented on src/main/java/server/managers/ServiceManager.java:
```diff
@@ -0,0 +1,30 @@
+//
+
+
+}
```
is this file obsolete?

limxuanyu127 replied:
oh yes its obsolete, forgot to delete

sith0008 commented on src/main/java/server/translator/Translator.java:
```diff
@@ -0,0 +1,279 @@
+        int offsetMin = 0;
+        System.out.println();
+
+        switch (offset){
+            case 1:
+                offsetMin = 30;
+                break;
+            case 2:
+                offsetMin = 60;
+                break;
+            case -1:
+                offsetMin = -30;
+                break;
+        }
+
```
why do we have this switch ah?

should it just be 

`int offset = r.offset * 30`

limxuanyu127 replied:
oh ya hor HAHA yes, will change that

sith0008 reviewed (approved):
didnt really look through the business logic, i think we can review again when we test and debug. 

ok to merge. but see comments

xqlim1996 committed:
Minor fix for Client Listener

limxuanyu127 committed:
cleanup

limxuanyu127 committed:
merging with main
